lib/css-tools.js

```diff
@@ -0,0 +1,131 @@
+'use strict';
+
+var csstree     = require('css-tree'),
+    stable      = require('stable'),
+    specificity = require('csso/lib/restructure/prepare/specificity'),
+    List        = require('css-tree/lib/utils/list');
+
+
+function flattenToSelectors(cssAst) {
+  var selectors   = [];
+
+  csstree.walkRules(cssAst, function(node) {
+    if (node.type !== 'Rule') {
+        return;
+    }
+
+    var atrule = this.atrule;
+    var rule   = node;
+
+    node.selector.children.each(function(selectorNode, selectorItem) {
+        var selector = {
+            item:    selectorItem,
+            atrule:  atrule,
+            rule:    rule,
+            pseudos: []
+        };
+
+        selectorNode.children.each(function(selectorChildNode, selectorChildItem, selectorChildList) {
+            if(selectorChildNode.type == 'PseudoClassSelector' || 
+               selectorChildNode.type == 'PseudoElementSelector') {
+                selector.pseudos.push({ item: selectorChildItem, list: selectorChildList });
+            }
+        });
+
+        selectors.push(selector);
+    });
+});
+
+  return selectors;
+}
+
+
+function filterByMqs(selectors, useMqs) {
+  return selectors.filter(function(selector) {
+    if(selector.atrule === null) {
+      return useMqs.indexOf('') > -1;
+    }
+
+    var mqName = selector.atrule.name;
+    var mqStr  = mqName;
+    if(selector.atrule.expression.type === 'MediaQueryList') {
+      var mqExpr = csstree.translate(selector.atrule.expression);
+          mqStr  = [mqName, mqExpr].join(' ');
+    }
+
+    return useMqs.indexOf(mqStr) > -1;
+  });
+}
+
+function filterByPseudos(selectors, usePseudos) {
+  return selectors.filter(function(selector) {
+    var pseudoNodes = [];
+    for(var pseudo of selector.pseudos) {
+      pseudoNodes.push(pseudo.item.data);
+    }
+    var pseudoSelectorsStr = csstree.translate({
+      type: 'Selector',
+      children: new List().fromArray(pseudoNodes)
```

Comment:
Ideally it would be possible to get the containing list of an item so just the items need to be passed, currently the list of an item has to be stored with it together for removing itself from the list.